src/v/cloud_storage/partition_manifest.cc

```diff
@@ -313,6 +437,49 @@ partition_manifest::get_insert_iterator() {
     return std::inserter(_segments, _segments.begin());
 }
 
+template<class Container>
+class insert_iterator_conditional {
+public:
+    template<class Cond>
+    insert_iterator_conditional(
+      Cond&& condition, Container& c, typename Container::iterator it)
+      : _iter(c, it)
+      , _cond(std::forward<Cond>(condition)) {}
+
+    insert_iterator_conditional&
+    operator=(const typename Container::value_type& value) {
+        if (_cond(value)) {
+            _iter = value;
+        }
+        return *this;
+    }
+
+    insert_iterator_conditional&
+    operator=(typename Container::value_type&& value) {
+        if (_cond(value)) {
+            std::insert_iterator<Container>::operator=(std::move(value));
+        }
+        return *this;
+    }
+
+    // No-op operators
+    insert_iterator_conditional& operator*() { return *this; }
+    insert_iterator_conditional& operator++() { return *this; }
+    const insert_iterator_conditional operator++(int) { return *this; }
+
+private:
+    std::insert_iterator<Container> _iter;
+    std::function<bool(const typename Container::value_type&)> _cond;
+};
+
+template<class Func, class Container>
+insert_iterator_conditional<Container> conditional_inserter(
+  Func&& condition, Container& c, typename Container::iterator it) {
+    insert_iterator_conditional<Container> result(
+      std::forward<Func>(condition), c, it);
+    return result;
+}
+
 partition_manifest
 partition_manifest::difference(const partition_manifest& remote_set) const {
```

Comment:
It's tempting but we're also using it in read replicas which has to download manifest and compute delta.

Reply:
Ah, ok. I don't know much about how read replicas work, but presumably a read replica just has to bring its local manifest to be equal to the remote one? So the mechanical difference on the `_segments` and `_replaced` sets should suffice?

Reply:
I actually removed difference method, right now read-replica sync downloads the manifest and adds all new segments that wasn't there previously. RR doesn't need to do any maintenance so it's ok not to have complete list of removed segments. But the code will have to be adjusted once we will implement compacted segment re-uploading.

Reply:
The `add_segments` method now still accepts the `partition_manifest` but it's used as a collection of segments. This is done to avoid including `cloud_storage/partition_manifest.h` in the `cluster/archival_metadata_stm.h` directly.